src/brogue/Combat.c

```diff
@@ -1670,17 +1670,12 @@ void killCreature(creature *decedent, boolean administrativeDeath) {
         } else {
             pmap[x][y].flags &= ~HAS_MONSTER;
         }
-        removeCreature(dormantMonsters, decedent);
-        removeCreature(monsters, decedent);
-
-        if (decedent->leader == &player
-            && !(decedent->info.flags & MONST_INANIMATE)
-            && (decedent->bookkeepingFlags & MB_HAS_SOUL)
-            && !administrativeDeath) {
-            prependCreature(&purgatory, decedent);
-        } else {
-            prependCreature(&graveyard, decedent);
 
+        // This must be done at the same time as removing the HAS_MONSTER flag, or game state might
+        // end up inconsistent.
+        decedent->bookkeepingFlags |= MB_HAS_DIED;
```

Comment:
What information does this provide over MB_IS_DYING? They seem to be set at very similar times

Reply:
I tried using only the existing flag but changing it to be set later, but that resulted in some explosions happening twice, as well as other desyncs; this is likely due to one of the function calls in `killCreature` that happens between the two flags being set and which probably behaves differently when `MB_IS_DYING` is set. And if I use `MB_IS_DYING` for everything but keep it where it currently is, that results in crashes when the dungeon-rendering code can't find a monster in the iterator because it's currently dying (since the iterator was changed to skip over dead monsters).

I'm not sure if it's possible to fix this without adding the new flag, at least not without a significant rewrite of the explosion-handling code, and that would almost certainly break recordings.